# Introduction to Machine Learning

Machine Learning (ML) is a field of computer science where we build systems that learn patterns from data instead of being explicitly programmed.

In ML, we provide data to an algorithm, and the algorithm discovers relationships between input variables (features) and the target output. This allows the model to make predictions on new, unseen data.

Below is a simple conceptual diagram of how machine learning works:



             ┌───────────────────┐
             │   Training Data   │
             │  (Inputs + Labels)│
             └─────────┬─────────┘
                       │
                       ▼
         ┌──────────────────────────┐
         │      ML Algorithm        │
         │ (Learns a relationship)  │
         └─────────┬────────────────┘
                   │
                   ▼
         ┌──────────────────────────┐
         │     Trained Model        │
         │ (Used for prediction)    │
         └─────────┬────────────────┘
                   │
                   ▼
          ┌──────────────────┐
          │   Predictions     │
          └──────────────────┘


In this notebook, we will introduce the main ideas of:

- Supervised vs Unsupervised Machine Learning  
- Classification  
- Regression  
- Classical machine learning algorithms used commonly in scikit-learn  

Each concept will be presented with simple diagrams and clear explanations.


# 🔵 Supervised vs Unsupervised Learning

Machine learning methods can be broadly divided into **supervised** and **unsupervised** learning, depending on whether we provide labeled outputs during training.

---

## 🔷 Supervised Learning

Supervised learning uses **labeled data**, meaning each training example includes both:

- **Input features (X)**
- **A known correct output (y)**

The goal is for the model to learn the mapping from X → y so it can make accurate predictions on new data.

### Diagram — Supervised Learning



         ┌─────────────────────────────┐
         │  Labeled Training Examples  │
         │  (Features + Correct Output)│
         └─────────────┬───────────────┘
                       │
                       ▼
             ┌───────────────────┐
             │   ML Algorithm    │
             │  Learns mapping   │
             └─────────┬─────────┘
                       │
                       ▼
            ┌────────────────────┐
            │   Trained Model    │
            │ Predict new labels │
            └────────────────────┘


### Examples of supervised problems:
- Predicting if an email is spam or not (classification)
- Predicting house prices from size and location (regression)
- Determining if a tumor is benign or malignant (classification)

---

## 🔷 Unsupervised Learning

Unsupervised learning uses **unlabeled data**. The model does not know the correct outputs.  
Instead, it tries to find **patterns, clusters, or structure** inside the data.

### Diagram — Unsupervised Learning



          ┌─────────────────────┐
          │   Unlabeled Data    │
          │   (Only features)   │
          └──────────┬──────────┘
                     │
                     ▼
        ┌──────────────────────────┐
        │   Unsupervised Algorithm │
        │  Finds structure/groups  │
        └──────────┬──────────────┘
                     │
                     ▼
         ┌─────────────────────────┐
         │   Patterns / Clusters   │
         │   or Lower Dimensions   │
         └─────────────────────────┘


# 🔵 Classification

**Classification** is a supervised machine learning task where the goal is to predict **categories** or **class labels**.

The output is **discrete**, meaning the model chooses from a finite set of possible classes such as:
- Spam / Not Spam  
- Positive / Negative  
- Dog / Cat  
- Digit 0–9  

Classification models learn boundaries that separate different categories based on the features.

---

## 🔷 Intuition Behind Classification

A classification model tries to learn **decision boundaries** that divide the feature space into regions representing different classes.

For example, with two features, the model might learn a boundary line (or curve) that separates Class A from Class B.

### Diagram — Classification Boundary (Concept)
<img src="classification.png" height="200">


In real problems, the boundaries can be much more complex depending on the algorithm used.

---

## 🔷 Classification Pipeline

Typical steps in a classification task include:

1. **Collect data** (features + labels)
2. **Split into training and testing sets**
3. **Train a classifier** (e.g., Logistic Regression, SVM, Decision Tree)
4. **Evaluate performance** (accuracy, confusion matrix, F1-score)
5. **Use the model to classify new data**
---

## 🔷 Examples of Classification Problems

- **Medical diagnosis:** benign vs malignant tumor  
- **Email filtering:** spam vs not spam  
- **Finance:** approve/deny a loan  
- **Computer vision:** identifying object categories in images  
- **Text processing:** classifying sentiment (positive/negative)

---

Classification is one of the fundamental problems in machine learning and a key component of many real-world applications.



# 🔵 Regression

**Regression** is a supervised machine learning task where the goal is to predict **continuous numerical values**.

The output is a real number, not a category.  
Regression models learn the relationship between input features and a numeric target variable.

---

## 🔷 What Regression Predicts

Examples of regression problems include:

- Predicting house prices  
- Predicting temperature  
- Predicting sales or revenue  
- Estimating a patient's risk score  
- Predicting exam scores  

Regression answers questions like **“How much?”** or **“What value?”**

---

## 🔷 Intuition Behind Regression

A regression model tries to fit a line, curve, or function that best describes the relationship between the inputs and the continuous output.

### Simple Linear Regression Concept

<img src="Regression.png" width="300">



In more complex scenarios, models may fit curves, surfaces, or nonlinear functions.

---

## 🔷 Regression Pipeline

Typical steps in a regression workflow:

1. **Collect data** (features + continuous target)
2. **Split into training/testing sets**
3. **Train a regression model**  
   (Linear Regression, Decision Tree Regressor, Random Forest Regressor, etc.)
4. **Evaluate model performance**
   - MSE, RMSE, MAE, R²
5. **Use the trained model to predict numeric values**
---

## 🔷 Common Metrics Used in Regression

- **MSE (Mean Squared Error)**  
- **RMSE (Root Mean Squared Error)**  
- **MAE (Mean Absolute Error)**  
- **R² Score (Coefficient of Determination)**

These metrics quantify how close the predicted values are to the true values.

---

Regression is essential for forecasting, modeling relationships, and many decision-making processes across science, engineering, economics, and healthcare.


# 🔵 Classical Machine Learning Algorithms

Classical machine learning algorithms are the foundation of many practical ML applications.  
They are fast, interpretable, and easy to use — especially through the `scikit-learn` library.

Below are the most commonly used algorithm families, grouped by type, with simple diagrams to visualize how they work conceptually.

## 1. Linear & Logistic Models

These models learn a **linear relationship** between the input features and the output.
<img src="Logistic-regression-and-linear-regression.png" width="800"> 

### 🔹 Linear Regression (for regression)
Fits a straight line or plane to minimize the prediction error.

### 🔹 Logistic Regression (for classification)
Despite the name, it is a **classifier**, not a regression model.  
It models the probability of belonging to a class.





## 2. Decision Trees

Decision trees split the data using simple rules:

- "Is feature X > 3.7?"
- "Is feature Y <= 12?"

They create a flowchart-like structure that leads to a prediction.

### Diagram — Decision Tree Structure

                   ┌───────────────┐
                   │  Feature X > 5?│
                   └───────┬───────┘
                           │
             ┌─────────────┴─────────────┐
             ▼                           ▼
     ┌──────────────┐           ┌──────────────────┐
     │ Predict Class│           │   Check Feature Y │
     │      0       │           │     (next split)  │
     └──────────────┘           └──────────┬────────┘
                                            │
                                      ┌─────┴───────┐
                                      ▼             ▼
                                Class 1         Class 0

                   ┌───────────────────┐
                   │   Feature X > 5?  │
                   └──────────┬────────┘
                              │
              ┌───────────────┴────────────────┐
              ▼                                 ▼
      ┌──────────────────┐             ┌──────────────────────┐
      │   Leaf Node       │             │     Next Split:      │
      │ Predict Class 0   │             │    Feature Y ≤ 3?    │
      └──────────────────┘             └──────────┬───────────┘
                                                  │
                                          ┌───────┴────────┐
                                          ▼                ▼
                                  Predict Value 12    Predict Value 20





                   ┌───────────────────┐
                   │   Feature X > 5?  │
                   └──────────┬────────┘
                              │
              ┌───────────────┴────────────────┐
              ▼                                 ▼
      ┌──────────────────┐             ┌──────────────────────┐
      │   Leaf Node       │             │     Next Split:      │
      │ Predict Class 0   │             │    Feature Y ≤ 3?    │
      └──────────────────┘             └──────────┬───────────┘
                                                  │
                                          ┌───────┴────────┐
                                          ▼                ▼
                                  Predict Value 12    Predict Value 20


## 4. Support Vector Machines (SVM)

Support Vector Machines are powerful and flexible models that can be used for **both classification and regression**:

- **SVC** (Support Vector Classification)
- **SVR** (Support Vector Regression)

SVMs work by finding a function or boundary that **maximizes the margin** — the distance between the decision boundary (or regression function) and the closest data points, called *support vectors*.

### SVMs can model:
- **Linear boundaries**
- **Nonlinear boundaries** using kernel functions (RBF, polynomial, sigmoid)

### Examples of tasks:
- **Classification:** spam detection, image recognition, medical diagnosis  
- **Regression:** predicting temperature, forecasting sales, continuous value prediction
<img src="SVM.png" width="400"> 


# Simple Machine Learning Code Demo with Scikit-Learn

In this notebook, you will find a practical step by step coding workflow using the Scikit-Learn library and its **California Housing** dataset. You will Learn about concepts like:
- Train and Test Split
- Baseline Model
- R2 Score
- Hyperparameter Tuning
- and more...

If at any point, you find this notebook difficult to understand, please refer to my step by step video tutorial on YouTube: https://youtu.be/-IvNzmrcyUM

## Load Dataset

In the following cell, we will load the California Housing dataset and split it into features and targets. We will also print some basic information about the size and content of the dataset.

In [1]:
from sklearn import datasets

# load dataset from sklearn
housing = datasets.fetch_california_housing()

# split data to features and targets
x = housing.data
y = housing.target 

print("DATASET COLUMN NAMES:\n", housing.feature_names)
print("-----------------------")
print("nuber of rows:", len(x))
print("nuber of columns:", len(housing.feature_names))

DATASET COLUMN NAMES:
 ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']
-----------------------
nuber of rows: 20640
nuber of columns: 8


## Understand Sample and Target Data

In the cell below, we will print the first sample and the first target of the dataset, explaining what the stored information represents.
<br>
<br>
The information below represents a group of population who's:

- **Median Income**: $83,252
- **House Age**: 41 years old
- **Average Rooms**: 7 rooms (approx.)
- **Average Bedrooms**: 1 bedroom (approx.)
- **Population**: 322 people in this group.
- **Average Occup**: 2.5 occupants per household (approx.)
- **Latitude**: houses located in latitude 37.88.
- **Longitude**: houses located in longitude -122.23.

Finally, the **target** represents the dollar value of the house: $452,600

In [2]:
print(x[0]) # sample
print(y[0]) # target

[   8.3252       41.            6.98412698    1.02380952  322.
    2.55555556   37.88       -122.23      ]
4.526


## Train and Test Split

In the cell below, we will split our data into training samples and test samples.
<br>
It's an important step, when we reserve some of the data for model evaluation purposes. It will help us determine if the model actually learned the data or memorised it.
<br>
And the idea is, once the model is done training on the train set, we must expose it to samples it has never seen before (test set) to determine how well it performs. That's why we usually reserve about a 1/5th of the data for testing, instead of using the entire data during training.

In [3]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x, 
    y, 
    test_size=0.2,
    random_state=432
)

print("number of train samples:", len(x_train))
print("number of test samples:", len(x_test))

number of train samples: 16512
number of test samples: 4128


## Train Model on Data Subset

In the cell below, we will choose a machine learning algorithm and we will train it on the California Housing Data 

In [4]:
from sklearn.linear_model import LinearRegression

# choose algorithm
model = LinearRegression()
# train model on the train set
model.fit(x_train, y_train)

LinearRegression()

## Evaluate Model on Data Subset
To understand how well-trained and intelligent our model is, we will need to test whether it learned or memorised the data (as mentioned above).
<br>
<br>
For this, we will make the model "guess" (or predict) the dollar value of samples it has never seen before (`x_test`). 
<br>
Once we get a prediction from our model, we'll compare it to the actual dollar value (given by `y_test`), which will produce a model score.
<br>
<br>
In our example, we will use the R2 metric to obtain a score, and since it's the first time we are evaluating this model, this score will represent our **Baseline**.

In [5]:
from sklearn.metrics import r2_score

# get the model to "guess" the dollar value of the houses in the test sample
y_pred = model.predict(x_test)
# compare what the model "guessed" with the actual value from y_test
r2 = r2_score(y_test, y_pred)

print("R2 Score:", r2)

R2 Score: 0.6080229586580339


### Optimize Algorithms
How do we know that LinearRegression is the best algorithm for our type of data? 
<br>
The answer is: **WE DON'T!**
<br>
Which means that we need to try several algorithms to determine that, in our case: 
LinearRegression, GradientBoostingRegresso, RandomForestRegressor.
<br>
Once we compair the R2 score for each model, we will know which algorithm works best!

In [6]:
#%%
from sklearn.ensemble import (
    GradientBoostingRegressor, 
    RandomForestRegressor
)

# initialize models
LR = LinearRegression()
GBR = GradientBoostingRegressor()
RFR = RandomForestRegressor(n_jobs=-1)

# LR
LR.fit(x_train, y_train)
y_pred = LR.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("MODEL:", 'LR')
print("R2 SCORE:", r2)
print("-------------")


# GBR
GBR.fit(x_train, y_train)
y_pred = GBR.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("MODEL:", 'GBR')
print("R2 SCORE:", r2)
print("-------------")

# RFR
RFR.fit(x_train, y_train)
y_pred = RFR.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("MODEL:", 'RFR')
print("R2 SCORE:", r2)
print("-------------")

MODEL: LR
R2 SCORE: 0.6080229586580339
-------------
MODEL: GBR
R2 SCORE: 0.7878126561044645
-------------
MODEL: RFR
R2 SCORE: 0.8144324082315159
-------------


We notice that the `RandomForestRegressor` and the `GradientBoostingRegressor` perform better than the baseline by more than 19 points!
<br>
Which means that they are a better fit than `LinearRegression`, and we should use one of them instead.
<br>
There's just one problem with these new algorithms - they take a relatively long time to process and they are very heavy on our computer. We will solve it in the cell below.

### Optimize Algorithm Speed
If we'd like to optimize the algorithm speed, rather than the R2 score, we can do the following:
- **RandomForestRegressor**: add `n_jobs=-1` argument to use the maximum available number of CPU cores for processing
- **GradientBoostingRegressor**: use optimized and more efficient version of this algorithm, `HistGradientBoostingRegressor`

Let's apply these in the cell below:

In [7]:
from sklearn.ensemble import HistGradientBoostingRegressor

# initialize speed optimized models
RFR = RandomForestRegressor(n_jobs=-1)

# iterate over models, training and evaluating each
RFR.fit(x_train, y_train)
y_pred = RFR.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("MODEL:", 'RFR')
print("R2 SCORE:", r2)
print("-------------")

MODEL: RFR
R2 SCORE: 0.8132083254316401
-------------


We see that now, our optimized `GradientBoostingRegressor` model, produces a better R2 score than before! 
<br>
<br>
With `HistGradientBoostingRegressor`, we are now 4 points above the `GradientBoostingRegressor`score from the previous cell, and 24 points above the baseline! So let's choose this as our best-performing model and let's use it in the rest of this notebook.

### Optimize Hyperparameters
The last optimization technique we will see in this notebook, is optimization of hyperparameters for the best-performing model.
<br>
We can try different combinations of model parameter values to improve our score further and further, such as: `max_iter`, `learning_rate`, `max_depth`, etc.
<br>
<br>
In the cell below, we will find which number of trees works best for the `HistGradientBoostingRegressor`, optimizing a parameter named `n_features`

In [8]:
for i in [100, 200, 300, 400, 500]:
    model = HistGradientBoostingRegressor(
        max_iter = i
    )
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    r2 = r2_score(y_test, y_pred)
    print("NUMBER OF TREES:", i)
    print("R2 SCORE:", r2)
    print("-------------")

NUMBER OF TREES: 100
R2 SCORE: 0.8357033753676242
-------------
NUMBER OF TREES: 200
R2 SCORE: 0.8434477761308892
-------------
NUMBER OF TREES: 300
R2 SCORE: 0.8459729056684533
-------------
NUMBER OF TREES: 400
R2 SCORE: 0.8428608677423277
-------------
NUMBER OF TREES: 500
R2 SCORE: 0.8408486283109835
-------------


We see that the best R2 score is given by `max_iter=400`, but that's not all! We can optimize more than one hyperparameter, so let's also introduce `learning_rate` into the miz with a nested for loop:

In [9]:
for j in [0.1, 0.05, 0.001]:
    for i in [200, 300, 400]:
        model = HistGradientBoostingRegressor(
            max_iter = i,
            learning_rate = j
        )
        model.fit(x_train, y_train)
        y_pred = model.predict(x_test)
        r2 = r2_score(y_test, y_pred)
        print("NUMBER OF TREES:", i)
        print("LEARNING RATE:", j)
        print("R2 SCORE:", r2)
        print("-------------")

NUMBER OF TREES: 200
LEARNING RATE: 0.1
R2 SCORE: 0.841533177755672
-------------
NUMBER OF TREES: 300
LEARNING RATE: 0.1
R2 SCORE: 0.8482611179472842
-------------
NUMBER OF TREES: 400
LEARNING RATE: 0.1
R2 SCORE: 0.849778746909608
-------------
NUMBER OF TREES: 200
LEARNING RATE: 0.05
R2 SCORE: 0.8367141433224132
-------------
NUMBER OF TREES: 300
LEARNING RATE: 0.05
R2 SCORE: 0.8425870485987956
-------------
NUMBER OF TREES: 400
LEARNING RATE: 0.05
R2 SCORE: 0.842985293284621
-------------
NUMBER OF TREES: 200
LEARNING RATE: 0.001
R2 SCORE: 0.21533054681581298
-------------
NUMBER OF TREES: 300
LEARNING RATE: 0.001
R2 SCORE: 0.301890446792871
-------------
NUMBER OF TREES: 400
LEARNING RATE: 0.001
R2 SCORE: 0.3701081153517073
-------------


We see that the best R2 score is `0.846` given by:
- `max_iter=400`
- `learning_rate=0.05`

So we should set it as our new best-performing model.
<br>
<br>
Please note, we've experimented only with 2 hyperparameters in this workflow, but in reality - you should experiment with as many hyperparameters combinations as possible!

## Model Saving
Great! So we found our best performing model, which is almost 25 points better than the baseline! Next, we will need to save this model to our system, so we never have to train it again.
<br>
<br>
For this we will use a library named `joblib`:

In [39]:
import joblib

model = HistGradientBoostingRegressor(
    max_iter=400, # best max iter
    learning_rate=0.05 # best learning rate
)
model.fit(x_train, y_train)

# save model
joblib.dump(model, "my_model.joblib")

['my_model.joblib']

### Ensure Model is Properly Saved
To make sure our model was properly saved, we will load (also with joblib) it and compare it to the existing model.

In [40]:
# evaluate existing model
y_pred = model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("existing model score:", r2)

# load saved model
saved_model = joblib.load("my_model.joblib")
# evaluate saved model
y_pred = saved_model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("saved model score:", r2)

existing model score: 0.8409061745888814
saved model score: 0.8409061745888814
